include/deal.II/base/std_cxx11/unique_ptr.h

```diff
@@ -0,0 +1,83 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2015 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE at
+// the top level of the deal.II distribution.
+//
+// ---------------------------------------------------------------------
+
+#ifndef __deal2__std_cxx11_unique_ptr_h
+#define __deal2__std_cxx11_unique_ptr_h
+
+
+#include <deal.II/base/config.h>
+
+#ifdef DEAL_II_WITH_CXX11
+
+#  include <memory>
+DEAL_II_NAMESPACE_OPEN
+namespace std_cxx11
+{
+  using std::unique_ptr;
+}
+DEAL_II_NAMESPACE_CLOSE
+
+#else
+
+#include <boost/shared_ptr.hpp>
+DEAL_II_NAMESPACE_OPEN
+namespace std_cxx11
+{
+  /**
+   * Implementation of a basic replacement for C++11's std::unique_ptr
+   * class.
+   *
+   * BOOST does not have a replacement for std::unique_ptr (because
+   * unique_ptr requires move semantics that aren't available unless
+   * you have a C++11 compiler -- in which case you also have
+   * std::unique_ptr; see for example
+   * http://stackoverflow.com/questions/2953530/unique-ptr-boost-equivalent)
+   *
+   * Consequently, we emulate the class by just wrapping a
+   * boost::shared_ptr in the cheapest possible way -- by just
+   * deriving from it and repeating the basic constructors. Everything
+   * else is inherited from the shared_ptr class.
+   *
```

Comment:
You can't easily make a "real" `unique_ptr` in C++03 because you can't emulate move semantics. Apparently boost 1.59 has something, but for the moment we only have 1.57 in bundled/. I thought it's not worth it -- by now, the majority of users probably already has C++11 compilers. We're also not using it in user interfaces at all.